### PR TITLE
Mobstate Refactor

### PR DESCRIPTION
**Pull request by Jezithyr**
## About the PR 
A complete refactor of mob state to utilize an event.

Additionally, mob state thresholds now have their own component and system. Eventually the mob threshold system will be depreciated in favour of body simulation systems. This also helps simplify the mobstate logic and differentiate functionality.

**_How is this useful?:_**
One of the major issues with the old mobstate system is that many different gameplay systems all interacted with it, this made it's logic very complex and resulted in alot of unneeded dependencies. This event allows multiple systems to interact with mob states without causing redundant dependencies.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## **!!!DOWNSTREAMS NOTICE!!!:**
There is a breaking change in YAML, move any current uses of the threshold node of 'MobState' in entity prototypes to a new component called 'MobThresholds', otherwise mob thresholds will not work properly. You will still need a MobState component! Specifying allowedStates in MobState is not required but prevents systems from setting the wrong state. I've already made this change in the WizDen YAML files if you need a reference.

**Changelog**

### CODE REVIEW

**Comment by PixelTheKermit** [comment]
Surgery, just a week away...

**Comment by Jezithyr** [comment]
This is required for wounds/bobbysim

File: Content.Shared/Mobs/Systems/MobStateSystem.StateMachine.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    protected virtual void OnEnterState(MobStateComponent component, MobState state)
+    {
+        OnStateEnteredSubscribers(component, state);
+    }
+
+    //Called right before stateChangedEvent gets raised
+    protected virtual void OnStateChanged(MobStateComponent component, MobState oldState, MobState newState) {}
+
+    //Called when exiting a state
+    protected virtual void OnExitState(MobStateComponent component, MobState state)
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Systems/MobStateSystem.StateMachine.cs:19]: Make em xmldocs

File: Content.Shared/Mobs/MobState.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        MobState OldMobState,
+        MobState NewMobState,
+        EntityUid? Origin = null)
+    {
+        this.Component = Component;
+        this.OldMobState = OldMobState;
+        this.NewMobState = NewMobState;
+        this.Origin = Origin;
+    }
+    public EntityUid Entity => Component.Owner;
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/MobState.cs:33]: Should just pass EntityUid and use that if we're killing .Owner

File: Content.Shared/Mobs/Systems/MobStateSystem.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        SubscribeLocalEvent<MobStateComponent, ComponentHandleState>(OnHandleComponentState);
+    }
+
+    private void OnHandleComponentState(EntityUid uid, MobStateComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not MobStateComponentState state)
+            return;
+
+        component.CurrentState = state.CurrentState;
+        component.AllowedStates = state.AllowedStates;
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Systems/MobStateSystem.cs:35]: Should copy allowedstates to a new dictionary like you did with mobthresholds

File: Content.Shared/Stunnable/SharedStunSystem.cs [hunk 4]
```diff
@@ -71,6 +75,38 @@ public override void Initialize()
             SubscribeLocalEvent<StunnedComponent, PickupAttemptEvent>(OnAttempt);
             SubscribeLocalEvent<StunnedComponent, IsEquippingAttemptEvent>(OnEquipAttempt);
             SubscribeLocalEvent<StunnedComponent, IsUnequippingAttemptEvent>(OnUnequipAttempt);
+            SubscribeLocalEvent<MobStateComponent, MobStateChangedEvent>(OnMobStateChanged);
+        }
+
+
+
+        private void OnMobStateChanged(EntityUid uid, MobStateComponent component, MobStateChangedEvent args)
+        {
```
Comment (metalgearsloth) [inline Content.Shared/Stunnable/SharedStunSystem.cs:84]:
```suggestion

        private void OnMobStateChanged(EntityUid uid, MobStateComponent component, MobStateChangedEvent args)
        {
```

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        VerifyThresholds(uid, thresholdComponent);
+    }
+
+    private void OnDamaged(EntityUid uid, MobThresholdsComponent mobThresholdsComponent, DamageChangedEvent args)
+    {
+        var mobStateComp = EnsureComp<MobStateComponent>(uid);
+        CheckThresholds(uid, mobStateComp, mobThresholdsComponent, args.Damageable);
+    }
+
+    //Call this if you are somehow change the amount of damage on damageable without triggering a damageChangedEvent
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:219]: xml

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    }
+
+    private bool EnsureComps(EntityUid target, [NotNullWhen(true)] MobStateComponent? mobStateComponent,
+        [NotNullWhen(true)] MobThresholdsComponent? thresholdComponent,
+        [NotNullWhen(true)] DamageableComponent? damageableComponent)
+    {
+        if (Resolve(target, ref thresholdComponent) && Resolve(target, ref mobStateComponent) &&
+            Resolve(target, ref damageableComponent))
+            return true;
+        Logger.Error("Target entity does not have damageable or mobThreshold components!");
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:271]: Should use ISawmill for logging.

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+        this.Entity = Entity;
+        this.ThresholdComp = ThresholdComp;
+        this.DamageableComp = DamageableComp;
+        this.MobStateComp = MobStateComp;
+    }
+
+    public EntityUid Entity { get; init; }
+    public MobStateComponent MobStateComp { get; init; }
+    public MobThresholdsComponent ThresholdComp { get; init; }
+    public DamageableComponent DamageableComp { get; init; }
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:329]: readonly record struct events my beloved.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Shouldn't need `this` in C#

File: Content.Shared/Mobs/MobState.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    public MobStateComponent Component { get; init; }
+    public MobState OldMobState { get; init; }
+    public MobState NewMobState { get; init; }
+    public EntityUid? Origin { get; init; }
+}
+public static class A
+{
+    //^.^
+}
```
Comment (AJCM-git) [inline Content.Shared/Mobs/MobState.cs:42]: A

File: Resources/Prototypes/Entities/Mobs/NPCs/revenant.yml [hunk 9]
```diff
@@ -62,7 +62,6 @@
     interfaces:
     - key: enum.StoreUiKey.Key
       type: StoreBoundUserInterface
-  - type: MobState
   - type: Visibility
```
Comment (AJCM-git) [inline Resources/Prototypes/Entities/Mobs/NPCs/revenant.yml:65]: No mob thresholds?

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Looks good, just need to document the public methods and fields and fix A

File: Content.Shared/Mobs/MobState.cs (hunk 8 above)
Comment (Jezithyr) [reply Content.Shared/Mobs/MobState.cs:42]: dammit you found it :(

File: Content.IntegrationTests/Tests/Commands/RejuvenateTest.cs [hunk 10]
```diff
@@ -22,7 +23,7 @@ public sealed class RejuvenateTest
   components:
   - type: Damageable
     damageContainer: Biological
-  - type: MobState
+  - type: MobThresholds
```
Comment (mirrorcult) [inline Content.IntegrationTests/Tests/Commands/RejuvenateTest.cs:26]: doesnt it also need mobstate

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+            ent1DeadThreshold = 0;
+
+        if (!TryGetThresholdForState(ent2, MobState.Dead, out var ent2DeadThreshold, threshold2))
+            ent2DeadThreshold = 0;
+
+        damage = (oldDamage.Damage / ent1DeadThreshold.Value) * ent2DeadThreshold.Value;
+        return true;
+    }
+
+    public void SetMobStateThreshold(EntityUid uid, FixedPoint2 damage, MobState mobState,
```
Comment (mirrorcult) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:203]: xmldoc

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+            _alerts.ClearAlertCategory(uid, AlertCategory.Health);
+    }
+
+    private void OnUpdateMobState(EntityUid uid, MobThresholdsComponent component, ref UpdateMobStateEvent args)
+    {
+        if (component.CurrentThresholdState != MobState.Invalid)
+            args.State = component.CurrentThresholdState;
+    }
+
+    public FixedPoint2 GetThresholdForState(EntityUid uid, MobState mobState,
```
Comment (mirrorcult) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:70]: same for all public methods in this class. should maybe #region the private/public apis apart from eachother

File: Content.Shared/Stunnable/SharedStunSystem.cs [hunk 13]
```diff
@@ -71,6 +75,38 @@ public override void Initialize()
...
+            if (!TryComp<StatusEffectsComponent>(uid, out var status))
+            {
+                return;
+            }
+            switch (args.NewMobState)
+            {
+                case MobState.Alive:
```
Comment (mirrorcult) [inline Content.Shared/Stunnable/SharedStunSystem.cs:91]: why is this necessary? if they died and were revived in the middle of a stun they should stay stunned imo. that or have a field on the statuseffectprototype for whether it should get removed on death/crit so this can be generalized

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+    {
+        args.State = new MobThresholdComponentState(component.CurrentThresholdState,
+            new Dictionary<FixedPoint2, MobState>(component.Thresholds));
+    }
+
+    private void MobThresholdMapInit(EntityUid uid, MobThresholdsComponent component, MapInitEvent args)
+    {
+        // TODO remove when body sim is implemented
+        EnsureComp<MobStateComponent>(uid);
+        EnsureComp<DamageableComponent>(uid);
```
Comment (mirrorcult) [inline Content.Shared/Mobs/Systems/MobThresholdSystem.cs:47]: these should just be on the protos instead of ensured in map init

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Shared/Stunnable/SharedStunSystem.cs (hunk 13 above)
Comment (Jezithyr) [reply Content.Shared/Stunnable/SharedStunSystem.cs:91]: That is what it is on master, and this is primarily a MobState refactor PR. I want to avoid changing anything mechanically

File: Content.Shared/Mobs/Systems/MobThresholdSystem.cs (hunk 14 above)
Comment (Jezithyr) [reply Content.Shared/Mobs/Systems/MobThresholdSystem.cs:47]: Yeah removing this. Originally I did this to minimize downstream YAML changes but @DrSmugleaf convinced me that this is an anti-pattern and it's better to just update the yamls instead.

File: Resources/Prototypes/Entities/Mobs/NPCs/revenant.yml (hunk 9 above)
Comment (Jezithyr) [reply Resources/Prototypes/Entities/Mobs/NPCs/revenant.yml:65]: Revenant cannot die, even on master. Their mobState component doesn't have any states in it. With the new system that can result in issues

**Comment by PixelTheKermit** [comment]
Surgery is just a week away.
